libraries/c_sdk/standard/mqtt/src/iot_mqtt_api.c

```diff
@@ -525,13 +471,19 @@ static void _destroyMqttConnection( _mqttConnection_t * pMqttConnection )
     IotMqtt_Assert( pMqttConnection->pingreqPacketSize == 0 );
 
     /* Remove all subscriptions. */
-    IotMutex_Lock( &( pMqttConnection->subscriptionMutex ) );
-    IotListDouble_RemoveAllMatches( &( pMqttConnection->subscriptionList ),
-                                    _mqttSubscription_setUnsubscribe,
-                                    NULL,
-                                    _mqttSubscription_tryDestroy,
-                                    offsetof( _mqttSubscription_t, link ) );
-    IotMutex_Unlock( &( pMqttConnection->subscriptionMutex ) );
+
+    contextIndex = _IotMqtt_getContextIndexFromConnection( pMqttConnection );
+
+    if( contextIndex >= 0 )
+    {
+        xSemaphoreTake( ( SemaphoreHandle_t ) &( connToContext[ contextIndex ].subscriptionMutex ), portMAX_DELAY );
+        IotMqtt_RemoveAllMatches( &( connToContext[ contextIndex ].subscriptionArray ),
+                                  _mqttSubscription_setUnsubscribe,
+                                  NULL,
+                                  _mqttSubscription_tryDestroy );
```

Comment:
Is this operation clearing all the subscription entries for the mqttConnection? Can you please add some more comments to point out what is happening here?